### PR TITLE
Implemented CookieModule

### PR DESCRIPTION
**Pull request by VladimirGlinskii**

### CODE REVIEW

File: common/src/cookie/cookie.module.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { CookieOptions } from './models';
+import { REQUEST, RESPONSE } from './tokens';
+
+@NgModule()
+export class CookieModule {
+  static forRoot(config?: CookieConfig): ModuleWithProviders<CookieModule> {
+    const providers: Array<Provider> = [
+      { provide: CookieOptions, useValue: config?.defaultOptions }
+    ];
+    if (config?.requestToken && config?.responseToken) {
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.module.ts:12]:
```suggestion

    if (config?.requestToken && config?.responseToken) {
```

File: common/src/cookie/cookie.service.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        const currentCookie = split.split('=');
+        cookies[decodeURIComponent(currentCookie[0])] = decodeURIComponent(currentCookie[1]);
+      }
+    }
+
+    return cookies;
+  }
+
+  public put(name: TKey, value: string, _options?: CookieOptions): void {
+    let cookieString = encodeURIComponent(name) + '=' + encodeURIComponent(value) + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:77]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    const options = this.getOptions(_options);
+
+    if (options.expires) {
+      cookieString += 'expires=' + options.expires.toUTCString() + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:81]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (options.maxAge) {
+      cookieString += 'Max-Age=' + options.maxAge + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:85]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (options.path) {
+      cookieString += 'path=' + options.path + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:89]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (options.domain) {
+      cookieString += 'domain=' + options.domain + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:93]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (options.secure) {
+      cookieString += 'secure;';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:97]: can we use `Template literals` here?

File: common/src/cookie/cookie.service.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (options.sameSite) {
+      cookieString += 'sameSite=' + options.sameSite + ';';
```
Comment (astorozhevsky) [inline common/src/cookie/cookie.service.ts:101]: can we use `Template literals` here?

**Review by astorozhevsky — CHANGES_REQUESTED** [review]

**Comment by astorozhevsky** [comment]
@ipakhomov @dmitryusenko could you please review?

File: common/src/cookie/cookie.module.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      providers.push(
+        { provide: REQUEST, useExisting: config.requestToken },
+        { provide: RESPONSE, useExisting: config.responseToken }
+      )
```
Comment (dmitryusenko) [inline common/src/cookie/cookie.module.ts:16]:
```suggestion
      );
```

File: common/src/cookie/cookie.service.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+  public get(key: TKey): string {
+    if (this.check(key)) {
+      const regExp: RegExp = this.getCookieRegExp(encodeURIComponent(key));
+      const result: RegExpExecArray = regExp.exec(this.cookieString);
+
+      return decodeURIComponent(result[1]);
+    }
+
+    return null;
```
Comment (dmitryusenko) [inline common/src/cookie/cookie.service.ts:46]:
```suggestion
    const regExp: RegExp = this.getCookieRegExp(encodeURIComponent(key));
    const result: RegExpExecArray = regExp.exec(this.cookieString);

    return (result !== null)
      ? decodeURIComponent(result[1])
      : null;
```

File: common/src/cookie/cookie.service.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  ) {
+    this.document = this.injector.get(DOCUMENT);
+    this.platformID = this.injector.get(PLATFORM_ID);
+    this.request = this.injector.get(REQUEST, undefined, InjectFlags.Optional);
+    this.response = this.injector.get(RESPONSE, undefined, InjectFlags.Optional);
+    this.defaultOptions = this.injector.get(CookieOptions, undefined, InjectFlags.Optional);
+    this.isDocumentAccessible = isPlatformBrowser(this.platformID);
+  }
+
+  public check(key: TKey): boolean {
```
Comment (dmitryusenko) [inline common/src/cookie/cookie.service.ts:32]: @VladimirGlinskii 

Maybe we don't need this method. If it's needed, then we could rename it to `has`.

File: common/src/cookie/cookie.service.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  public getAll(): Record<string, string> {
+    const cookies: Record<string, string> = {};
+
+    if (this.cookieString?.length) {
+      const splits = this.cookieString.split(/; ?/);
+      for (const split of splits) {
+        const currentCookie = split.split('=');
+        cookies[decodeURIComponent(currentCookie[0])] = decodeURIComponent(currentCookie[1]);
```
Comment (dmitryusenko) [inline common/src/cookie/cookie.service.ts:69]:
```suggestion
        const [name, value] = split.split('=');
        cookies[decodeURIComponent(name)] = decodeURIComponent(value);
```

File: common/src/cookie/cookie.service.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+      const regExp: RegExp = this.getCookieRegExp(encodeURIComponent(key));
+      const result: RegExpExecArray = regExp.exec(this.cookieString);
+
+      return decodeURIComponent(result[1]);
+    }
+
+    return null;
+  }
+
+  public getObject(name: TKey): object | null {
```
Comment (dmitryusenko) [inline common/src/cookie/cookie.service.ts:49]: @VladimirGlinskii 

We should unify the `TKey` argument name. In some methods it is named `key` and in other methods it's called `name`.

**Comment by VladimirGlinskii** [comment]
@astorozhevsky, @dmitryusenko Fixed your remarks.

**Review by astorozhevsky — APPROVED** [review]

**Review by ipakhomov — APPROVED** [review]
Looks good to me!

**Pull request merged**